### PR TITLE
Removing a copy on appending log to Persistent<T>.

### PR DESCRIPTION
**Pull request by songweijia**
We can avoid the copy when we appending to the persistent log. The idea is to use a 'data generator' lambda to do the work so that the object can be serialized directly to the persistent data region. This pull request involves changed `IDeltaSupport` API:
```
/** Removed in this pull request.
void IDeltaSupport::finalizeCurrentDelta(const DeltaFinalizer&);
**/

/** Added in this pull request. */
size_t IDeltaSupport::currentDeltaToBytes(uint8_t* const);
size_t IDeltaSupport::currentDeltaSize();
```

### CODE REVIEW

**Comment by songweijia** [comment]
A new configuration header file "derecho/config.h" is used to control the features during compilation. Also, change 'include' directives to <> for all headers not from a relative path.

**Comment by etremel** [comment]
I saw your e-mail and looked at the new commit that adds the "config.h" header and the "ENABLE_HMEM" option. This looks like a good idea and is an important fix to allow Derecho to be tested with the Libfabrics TCP provider. However, your commit also re-introduces a bug I fixed a while ago: include paths to Derecho headers within Derecho should use quotes, not <> brackets. As I documented in commit 90b587d16e96c9e86db9221b63022aa0cc99f999, using <> brackets for internal headers causes the compiler to search the system includes before the local source directory it is compiling, which makes it hard to rebuild Derecho after making changes (because the compiler will prefer the old "installed" headers instead of the changed ones in the source tree). 

Also, since the Derecho 2.4.0 release is kind of broken without the ENABLE_HMEM option, but the other changes in this pull request change the `Persistent<T>` IDeltaSupport API, I think we should separate the config changes into their own pull request and merge those first. That way we can create a release that is "old IDeltaSupport API but with TCP fixed", and another one that is "new IDeltaSupport API". I created a branch with just the config changes, called [`fix_tcp_provider`](https://github.com/Derecho-Project/derecho/tree/fix_tcp_provider), by copying all of your changes from the recent commit except the ones that changed quotes to <>. If you agree with this approach, we can create another pull request from that branch.

**Comment by songweijia** [comment]
Thank you for the input. 

First of all, I think that merging the config.h along with the DeltaSupport is not an issue: there only four places used that API, three in Derecho test, the other is Cascade. I've tested the persistent test and Cascade implementation. You have tested it in your signature feature. I believe it's better to include them instead of delaying it. 

About using <> or "" in the include directive: according to the GCC document [here](https://gcc.gnu.org/onlinedocs/gcc-2.95.3/cpp_1.html#SEC6), using "" will check the 'current directory', which is the directory of the current input file. I also did experiment to confirm this. Just cite the text below. 

#include "file"
This variant is used for header files of your own program. It searches for a file named file first in the **current directory**, then in the same directories used for system header files. The current directory is the directory of the current input file. It is tried first because it is presumed to be the location of the files that the current input file refers to. (If the `-I-' option is used, the special treatment of the current directory is inhibited.)

If you check my fix, you will notice that I didn't check all "" to <>. I only changed those does not make sense. For example
in rpc_manager.hpp, I kept `#include "../view.hpp"` as is but changed `#include "derecho/utils/logger.hpp"` to `#include <derecho/utils/logger.hpp>`. Why? because we can find `../view.hpp` from the **current directory** containing rpc_manager.hpp, but we cannot find `derecho/utils/logger.hpp` from the **current directory**!!! The behaviour will be falling back to "<>". Do you remember our discussion on why the fix in commit 90b587d does not work? Currently we just [remove the installed headers](https://github.com/Derecho-Project/derecho/blob/97e260cbc0728e2bba0f411f59927c2e6ea63e3b/scripts/build/build.sh#L83-L86) before building. This is same reason why I decide to drop this [pull request](https://github.com/Derecho-Project/cascade/pull/47) in cascade.

Actually, I found that the misusage is in the source code too.  I will add that to this pull request.

```
diff --git a/include/derecho/core/detail/rpc_manager.hpp b/include/derecho/core/detail/rpc_manager.hpp
index 2f338bbe..7f6cb8d6 100644
--- a/include/derecho/core/detail/rpc_manager.hpp
+++ b/include/derecho/core/detail/rpc_manager.hpp
@@ -6,11 +6,12 @@
 
 #pragma once
 
+#include <derecho/config.h>
 #include "../derecho_type_definitions.hpp"
 #include "../view.hpp"
-#include "derecho/mutils-serialization/SerializationSupport.hpp"
-#include "derecho/persistent/Persistent.hpp"
-#include "derecho/utils/logger.hpp"
+#include <derecho/mutils-serialization/SerializationSupport.hpp>
+#include <derecho/persistent/Persistent.hpp>
+#include <derecho/utils/logger.hpp>
 #include "derecho_internal.hpp"
 #include "p2p_connection_manager.hpp"
 #include "remote_invocable.hpp"

```

Sorry for the long response.

**Comment by songweijia** [comment]
By the way, how about [removing the current 2.4.0](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository) and re-release it with the included header style and config.h.in fixed?

**Comment by songweijia** [comment]
I also changed `#include ""` to `#include <>` in the source code.

**Comment by songweijia** [comment]
Great! Actually I experimented again with the '<>' style of include. And I didn't see that problem anymore... In my experiment, gcc/g++ does behave exactly as the document said. Btw, it is important to point out that gcc/g++ preprocessor search for include directories in the following order:
1) current directory (for "" style of include)
2) path passed to -I options
3) path in CPLUS_INCLUDE_PATH/C_INCLUDE_PATH
4) The default system include paths like '/usr/include', or compiler headers

We can use gcc's -v option to check that.

I don't have a good explanation for the previous issue yet. But let's keep an eye on it and keep improving the regression test.

File: src/applications/tests/unit_tests/signed_log_test.cpp [hunk 1]
```diff
@@ -76,18 +76,22 @@ std::string StringWithDelta::get_current_state() const {
     return current_state;
 }
 
-void StringWithDelta::finalizeCurrentDelta(const persistent::DeltaFinalizer& finalizer) {
-    if(delta.size() == 0) {
-        dbg_default_trace("StringWithDelta: Calling finalizer with null buffer");
-        finalizer(nullptr, 0);
+size_t StringWithDelta::currentDeltaSize() {
+    return delta.size();
```
Comment (etremel) [inline src/applications/tests/unit_tests/signed_log_test.cpp:80]: Shouldn't this be mutils::bytes_size(delta)? The serialized size of a std::string includes an extra int for the length of the string, so bytes_size will be slightly bigger than string::size().

**Pull request closed without merging**